### PR TITLE
Update kernel signature to use kwargs, migrate all rules

### PR DESCRIPTION
**Pull request by juliuskunze**
~Has a remaining bug that I will still fix tomorrow.~

~@j-towns: Another bug is the following: Whenever the output slice is empty, the core crashes (reproduce with disabled tests).~

**Update: Both are fixed now.**

I updated `lazy_lax_test.py` directly for a better diff and left `delayed_lax_test.py` untouched for now, we should replace that after merging.

### CODE REVIEW

**Comment by j-towns** [comment]
This looks absolutely brilliant Julius! Looking forward to reviewing tomorrow. Will also take a look at the empty out slice bug.

File: fastar/rules.py [hunk 1]
```diff
@@ -23,10 +23,6 @@ def naryop_dependency_rule(prim, outstart, outcount, *operands, **params):
               for o, b, (_, inshape) in zip(operands, bdcast, inboxes)]
   return inboxes, incounts, (tuple(params.items()), None)
```
Comment (j-towns) [inline fastar/rules.py]: 🤔 how come the static meta returned here isn't changed? Don't you have to replace `tuple(params.items())` with `params`?

File: fastar/rules.py [hunk 2]
```diff
@@ -99,7 +95,7 @@ def naryop_kernel(prim, meta_static, meta_dyn, *operands):
 
 for op in naryops:
   dependency_rules[op] = naryop_dependency_rule(op)
-  kernels[op] = naryop_kernel(op)
+  kernels[op] = op.bind
```
Comment (j-towns) [inline fastar/rules.py:98]: Satisfyingly simple.

File: fastar/rules.py [hunk 3]
```diff
@@ -188,17 +174,17 @@ def slice_dependency_rule(
   out_shape = np.asarray(outcount.shape)
   if strides is None:
     inbox = np.add(start_indices, outstart), out_shape
-    return [inbox], [np.ones(inbox[1], int)], (None, None)
+    return [inbox], [np.ones(inbox[1], int)], (dict(strides=tuple([1] * operand.ndim)), None)
```
Comment (j-towns) [inline fastar/rules.py]: `operand.ndim * (1,)` is shorter than
`tuple([1] * operand.ndim)`

File: fastar/rules.py [hunk 4]
```diff
@@ -273,34 +252,40 @@ def dot_general_dependency_rule(
...
   padcount = outcount.size - insize
-  def outslice(inslice, padding_value):
-    assert inslice is None or np.array_equal(inslice.shape, inshape)
-    return (lax.pad(inslice, padding_value,
-                    zip(offset, np.array(outcount.shape) - limit, interior))
-            if insize else jnp.full(outcount.shape, padding_value, operand.dtype))
   return ([(instart, inshape) if insize else None, ([], [])],
-          [incount, padcount], outslice)
+          [incount, padcount], (dict(
+    padding_config=tuple(zip(tuple(offset), tuple(np.array(outcount.shape) - limit), tuple(interior))),
```
Comment (j-towns) [inline fastar/rules.py]: I think the `tuple`s inside the zip won't do anything (possible I'm mistaken here so apologies if so), but I think
```python
zip(tuple(offset), tuple(np.array(outcount.shape) - limit), tuple(interior))
```
would be identical to
```python
zip(offset, np.array(outcount.shape) - limit, interior)
```

File: fastar/rules.py [hunk 5]
```diff
@@ -359,22 +342,31 @@ def conv_general_dilated_dependency_rule(
...
   return ([l_box, r_box], [lhs_count, rhs_count],
-          lambda lhs_slice, rhs_slice: out_transpose(conv(
-            lhs_pad(None if lhs_slice is None else lhs_transpose(lhs_slice),
-                    np.zeros((), lhs.dtype)),
-            rhs_transpose(rhs_slice))))
+          (dict(window_strides=window_strides, dimension_numbers=dimension_numbers, precision=precision, **lhs_pad_kwargs), None))
+
+def conv_general_dilated_kernel(lhs_slice, rhs_slice, window_strides, dimension_numbers, precision, **lhs_pad_kwargs):
+  lhs_spec, rhs_spec, out_spec = dimension_numbers
+  # abstract evaluations to retrieve shapes:
```
Comment (j-towns) [inline fastar/rules.py]: Is there an abstract eval here? Just wondering if this comment might be out of date.

File: fastar/test_util.py [hunk 6]
```diff
@@ -27,8 +27,11 @@ def check(thunk, atol=None, rtol=None):
   jtu.check_close(result, expected, atol, rtol)
 
 def random_box(rng, shape):
-  box_start = rng.randint(shape)
-  box_shape = rng.randint(1, np.subtract(shape, box_start - 1))
+  if np.any(np.less(shape, 0)):
+    raise ValueError
+  box_start = np.array([0 if s == 0 else rng.randint(s) for s in shape])
+  box_shape = np.array([0 if s == 0 else rng.randint(1, s - (start - 1))
+                        for start, s in zip(box_start, shape)])
```
Comment (j-towns) [inline fastar/test_util.py]: Same for this, you could probably express this in NumPy but I don't mind the explicit loop.

File: fastar/test_util.py [hunk 7]
```diff
@@ -27,8 +27,11 @@ def check(thunk, atol=None, rtol=None):
   jtu.check_close(result, expected, atol, rtol)
 
 def random_box(rng, shape):
-  box_start = rng.randint(shape)
-  box_shape = rng.randint(1, np.subtract(shape, box_start - 1))
+  if np.any(np.less(shape, 0)):
+    raise ValueError
+  box_start = np.array([0 if s == 0 else rng.randint(s) for s in shape])
```
Comment (j-towns) [inline fastar/test_util.py]: You could use a more terse NumPy-ish expression here, like
```python
box_start = rng.randint(np.maximum(s, 1))
```
I don't have a strong opinion about this though.

File: tests/lazy_lax_test.py [hunk 8]
```diff
@@ -146,11 +146,11 @@ def op_record(op, nargs, dtypes, rng_factory, tol=None):
 def test_nary(op_name, rng_factory, shapes, dtype, tol):
   rng = rng_factory(np.random)
   args = [rng(shape, dtype) for shape in shapes]
-  tu.check_lazy_fun(getattr(lax, op_name), *args, atol=tol, rtol=tol)
+  tu.check_multibox(lambda: getattr(lax, op_name)(*args), atol=tol, rtol=tol)
 
 LAX_REDUCE_OPS = [
   op_record("_reduce_sum", 1, number_dtypes, jtu.rand_default),
-  op_record("_reduce_prod", 1, number_dtypes, jtu.rand_small_positive),
+  # TODO op_record("_reduce_prod", 1, number_dtypes, jtu.rand_small_positive),
```
Comment (j-towns) [inline tests/lazy_lax_test.py]: Can you copy and paste [the new reducer tests](https://github.com/j-towns/fastar/blob/ed63d53cad68f87e4ebee2ecb91b195598442229/tests/delayed_lax_test.py#L152-L178) from delayed_lax_test.py? They fix this issue with reduce_prod and are probably more future-proof because they don't use any of the other private reducer functions.

File: fastar/rules.py (hunk 1 above)
Comment (j-towns) [reply fastar/rules.py]: I see now you haven't handled empty dicts yet, maybe this is the 'remaining bug' that you mentioned. See my [other comment](https://github.com/j-towns/fastar/pull/25#discussion_r554386609)...

File: fastar/core.py [hunk 9]
```diff
@@ -130,7 +130,7 @@ def _stack(*args):
           else jnp.stack([sup if a is null else a for a in args]))
 
 def compress(updates):
-  static = [u and (u.meta_static, u.inshapes) for u in updates]
+  static = [u and (u.meta_static and tuple(sorted(u.meta_static.items())), u.inshapes) for u in updates]
```
Comment (j-towns) [inline fastar/core.py]: I think you need to change
```python
u.meta_static and tuple(sorted(u.meta_static.items()))
```
to
```python
None if u.meta_static is None else tuple(sorted(u.meta_static.items()))
```
so that empty dicts don't slip through.

For safety you could add the line
```python
assert u is None or u.meta_static is None or isinstance(u.meta_static, dict)
```
just above.

**Review by j-towns — CHANGES_REQUESTED** [review]
Awesome work!

File: fastar/rules.py (hunk 3 above)
Comment (juliuskunze) [reply fastar/rules.py]: Good point! (Weird, somehow I thought this only works for lists.)

File: fastar/test_util.py (hunk 6 above)
Comment (juliuskunze) [reply fastar/test_util.py]: `rng.randint(0)` fails, therefore not vectorized

File: fastar/test_util.py (hunk 7 above)
Comment (juliuskunze) [reply fastar/test_util.py]: Good idea, done. Also added testing for empty boxes with non-zero shapes.

File: fastar/rules.py (hunk 1 above)
Comment (juliuskunze) [reply fastar/rules.py]: Yes, that was the bug. Thanks (:

File: fastar/core.py [hunk 10]
```diff
@@ -84,10 +84,9 @@ def toposort(haxpr):
   while childless_boxes:
     for e_sorted, e in reversed(zip(sorted_updates, haxpr.eqns)):
       o, = e.outvars
-      if o in childless_boxes:
-        obox = childless_boxes[o].pop()
-        if not childless_boxes[o]:
-          childless_boxes.pop(o)
+      oboxes = childless_boxes.get(o)
+      if oboxes:
+        obox = oboxes.pop()
```
Comment (j-towns) [inline fastar/core.py]: Need to be careful with changing this logic. I think the way it is in your proposed change could result in all of the lists  having a `None` at the end - once each list in childless boxes is empty there can still be a loop through the haxpr where every list would have a `None` added to it.

I think the right way to fix the bug you're hitting is to prevent empty lists being added to the dict [here](https://github.com/j-towns/fastar/blob/c97772f4d25154daf41530c508a1d141a7994183/fastar/core.py#L82-L83), i.e. you could first compute the `static_box_finder`, then assign it the `childless_boxes` dict only if the list is non-empty. Does that make sense?
